### ISSUE TITLE
Should there be a way for an individual task to break out of its portion of a forall?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#21742]
This issue is similar to, but different from requests we've had in the past for eurekas (e.g., https://github.com/chapel-lang/chapel/issues/12700). 

While writing some code recently, I found myself wondering if there should be a way for a task to break out of its portion of a forall.  That is, say that we were looking for a given value using a forall loop:

```chapel
var foundIt: bool
forall i in A.domain with (max reduce foundIt) {
  if (A[i] == val) {
    foundIt = true;
    // as a task, I arguably don't need to do any more iterations here, but have no alternative
  }
}
```

Even without a full eureka concept, work could be saved if, upon finding the value, a given task could skip the rest of its iterations.  That is, given a typical rewrite of a forall into something like:

```chapel
coforall tid in 0..<numTasks with (max reduce foundIt) {
  for i in myPortionOf(A.domain) {
    if (A[i] == val) {
      foundIt = true;
      break;  // here's a way I could skip the rest of my iterations without stopping everyone else as well
    }
  }
}
```

In the case where an array only had a single copy of the value, this wouldn't result in  much savings since the task in question would have to wait for everyone else to complete anyway.  But in the case where the array had many copies of the value (say every element of the array contained the value), it could reduce O(n) work to O(numTasks) work.

This seems like a reasonable concept, though how to express it is unclear.  Vanilla `break;` as proposed in #12700 seems more appropriate as a "break everyone out of this loop" concept.  Would we want to express it in some way that is specific to the task like `thisTask.break;` or `me.break;` or `Task.break();` or ...

Though I don't have a use case for it at present, a similar task-specific continue seems like it would be helpful as well.

**Comment by ronawho** [comment]
I have wanted something like this for checking if an array is sorted. If it's a fairly random distribution, it's likely that the tasks will independently discover that pretty quickly and could skip the rest of their iterations, resulting in a much quicker result without adding coordination overhead of a real eureka.

Rewriting Arkouda's [isSorted](https://github.com/Bears-R-Us/arkouda/blob/8c8c2378e3baf099b2ef7bd528121b2947c71eed/src/AryUtil.chpl#L65-L74) into the manual coforall is possible since it only supports a few distributions, but it'd be ideal to just be able to take advantage of the leader/follower iterators instead of manually chunking.

**Comment by benharsh** [comment]
Do we want users to be thinking about tasks when using a ``forall``? To me this seems more about coordinating with the iterator's logic than anything task-related.

How would this interact with a ``dynamic``-style iterator, where units of work are being continuously dealt out? At first glance it seems like this functionality would end one task, leaving the others to pick up more dynamically-dealt units of work.

It might shed some light on possible language mechanisms if we were to consider this from the perspective of providing this kind of behavior for ``begin`` statements. I think ``begin``s could also benefit from a way to basically "retire" the current task.

**Comment by bradcray** [comment]
> Do we want users to be thinking about tasks when using a forall?

I'd say we arguably already do through task-intents on foralls (e.g., `with (var i: int, + reduce total)`) and that when we tried to avoid such features, users just clamored for them.

> To me this seems more about coordinating with the iterator's logic than anything task-related.

That may be true; I was admittedly thinking only of our common-case iterators.

**Comment by e-kayrakli** [comment]
> Do we want users to be thinking about tasks when using a forall? To me this seems more about coordinating with the iterator's logic than anything task-related.

I agree with this. That's why I think this is related to https://github.com/chapel-lang/chapel/issues/16405 and features associated with it that we're planning to add to `forall`s. eg, allocating a per-locale array, barriering local tasks only, querying about number of tasks and where you're among them etc.

**Comment by bradcray** [comment]
Ben's comment about coordinating with the iterator got me to thinking that a hitch in my original desire for this (in the case of completely black-box parallel iterators) is that my ability to break is dependent on knowing that the loop I'm breaking out of is monotonically ascending.  Specifically, I wanted to assume that—having found the value—all subsequent iterations that I'll see will have indices that are strictly greater than this one.

That doesn't reduce my desire for the capability, but it does complicate it in terms of the interaction with the iterator.  In Engin's query framework, it might suggest that I'd want/need to be able to query whether the loop I was running through was increasing or decreasing or couldn't be depended upon to be in any order.

Also, opening up the isSorted case that Elliot shows to see if it has a similar issue (I think it doesn't), I'll note that my case could also have been written as a short-circuiting reduction (|| reduce foundIt), and that perhaps it's the short-circuiting nature of those operations that should trigger the ability for the early exit rather than anything manual in the loop body itself (?).  Although in my real case, I'm also doing a min reduce on the location where the element was found, and I'm imagining that would prevent the compiler from knowing (easily) that it could skip iterations...
